### PR TITLE
feat: add oidc support

### LINKED ISSUE #1090: FR: Generic OAuth OIDC option for SSO
### Feature Description

Instead of implementing every OAuth provider under the sun, support standard OpenID Connect (OIDC). Most brand name providers could actually be implemented via the exact same integration giving an admin freedom to implement a provider you don't want to provide on the flagship instance.

### Use Case

An organization implementing esigning on-premise is likely to have an identity provider in use and the diversity in the field is quite vast. Providing a well documented standard generic option for defining SSO gives them a way to plug it in on their own. OIDC is a modern alternative for SAML and well supported in IDMs. It's also significantly easier to configure for everyone involved than SAML.

### Proposed Solution

Utilize the NextAuth OAuth provider, allowing plugging in required config values via ENV variables. To simplify, only support the "well-known" autoconfig path. (Well supported by most identity providers, even open source ones.)

The parameters that need to be collected from ENV variables are:
- OIDC endpoint, i.e. the parameter to provide the next-auth `wellKnown` host part.
- Client ID, i.e. the `clientId` parameter.
- Client secret, i.e. the `clientSecret` parameter.
- Provider name, i.e. how do to label the button for selecting it.

Perhaps optional to be configurable, since there are reasonable defaults:
- Checks. For example `state` & `pkce` should be a reasonable default. If PKCE is not default enabled, an option needs to exist to enable it.
- Scopes. `openid email profile` should be a reasonable default.

### Alternatives (optional)

_No response_

### Additional Context

- next-auth oauth connector: https://github.com/nextauthjs/next-auth/blob/v4/packages/next-auth/src/providers/oauth.ts
- Documentation: https://next-auth.js.org/configuration/providers/oauth#using-a-custom-provider

### Please check the boxes that apply to this feature request.

- [X] I have searched the existing feature requests to make sure this is not a duplicate.
- [X] I have provided a detailed description of the requested feature.
- [X] I have explained the use case or scenario for this feature.
- [X] I have included any relevant technical details or design suggestions.
- [X] I understand that this is a suggestion and that there is no guarantee of implementation.
- [ ] I want to work on creating a PR for this issue if approved

### PR DESCRIPTION
**Pull request by tankerkiller125**
---
name: Pull Request
about: Submit changes to the project for review and inclusion
---

## Description

This PR adds generic OIDC as an authentication provider. This allows personal users and companies potentially to define whatever IdP they want as long as it supports the OIDC well known format. (Azure, Zitadel, Authentik, KeyCloak, Google, etc. all support it)

## Related Issue

Fixes #1090 

## Changes Made

- Adds OIDC buttons to the signin and registration pages
- Adds appropriate environment variables
- Adds migration to add OIDC to the `IdentityProvider` Enum

#### Zitadel
- Created application in Zitadel as an web app, with Client auth
- Enabled `User Info inside ID Token` in Token settings
- Copied client id and client secret to the new .ENV variables
- Copied the well-known URL from the URLs section to .ENV
- Created new account with OIDC provider button
- Verified email manually
- Signed into account with OIDC provider
- Logged out
- Signed into accounting again with OIDC provider

#### Authentik
- Created application in Authentik
- Copied client id and client secret to the new .ENV variables
- Copied the well-known URL from the URLs section to .ENV
- Created new account with OIDC provider button
- Verified email manually
- Signed into account with OIDC provider
- Logged out
- Signed into accounting again with OIDC provider

#### Azure AD
- Created application in Azure AD using OAuth2
- Copied client id and client secret to the new .ENV variables
- Copied the well-known URL from the URLs section to .ENV
- Created new account with OIDC provider button
- Verified email manually
- Signed into account with OIDC provider
- Logged out
- Signed into accounting again with OIDC provider

## Checklist

- [x] I have tested these changes locally and they work as expected.
- [ ] I have added/updated tests that prove the effectiveness of these changes.
- [x] I have updated the documentation to reflect these changes, if applicable.
- [x] I have followed the project's coding style guidelines.
- [ ] I have addressed the code review feedback from the previous submission, if applicable.

## Summary by CodeRabbit

- **New Features**
  - Introduced OIDC Single Sign-On (SSO) capabilities for both sign-in and sign-up processes, enhancing authentication options alongside existing Google SSO.
- **Enhancements**
  - Updated sign-in and sign-up forms to support OIDC SSO, including new OIDC sign-in/sign-up buttons.
- **Documentation**
  - Added necessary environment variables for OIDC configuration to the example file for clarity and setup guidance.

### CODE REVIEW

File: apps/web/src/components/forms/signup.tsx [hunk 1]
```diff
@@ -121,6 +127,19 @@ export const SignUpForm = ({ className, initialEmail, isGoogleSSOEnabled }: Sign
...
+      await signIn('google', { callbackUrl: SIGN_UP_REDIRECT_PATH });
+    } catch (err) {
+      toast({
+        title: 'An unknown error occurred',
+        description:
+          'We encountered an unknown error while attempting to sign you Up. Please try again later.',
+        variant: 'destructive',
+      });
+    }
+  };
```
[the remark this answers is not in the document]
Comment (tankerkiller125) [reply apps/web/src/components/forms/signup.tsx]: This has been fixed

**Comment by ElTimuro** [comment]
@tankerkiller125 this looks pretty neat 👀 Please give us some time to review this :)

**Comment by tankerkiller125** [comment]
@ElTimuro No problem, one thing that we might want to do before merging (that I didn't do) is add an environment variable to allow custom naming of the Buttons (so instead of saying Sign up with OIDC it could say something like "Sign up with Example Company" or something. I simply wasn't sure how to best handle exposing environment variables to the UI as I was initially making the PR.

**Comment by ElTimuro** [comment]
- Good point with the label
- though the default should be the current imo

**Comment by tankerkiller125** [comment]
I agree that the default should indeed be OIDC, but a customization option would be good.

**Comment by celevra** [comment]
any progress on this?

**Comment by ElTimuro** [comment]
- @celevra sadly not yet, since we are swamped
- But I'll bring it up again, since this is pretty encapsulated

**Comment by Mythie** [comment]
Hey there, apologies on the delay for this one.

It seems pretty neat but would only support one OIDC provider per deployment (instance) correct?

Ideally we would allow a given user to configure their OIDC provider and they would receive a URL for sign in or a subdomain.

That said I'd be happy to merge this as a first cut for self-hosters since it does seem really neat and then work toward the improvement in the future 😄

**Comment by tankerkiller125** [comment]
@Mythie Yes, this only support a single OIDC provider, I did think about the multi-OIDC provider thing, however I just didn't see a good way to pull it off with my limited knowledge of the project, and NodeJS itself (and the supporting libraries).

**Comment by leoc** [comment]
We would also need OIDC for Auth via Keycloak.

Can I do something to support this PR being merged?

**Comment by celevra** [comment]
> We would also need OIDC for Auth via Keycloak.
> 
> Can I do something to support this PR being merged?

Oidc is supported by keycloak, keycloak is also mentioned in the description of this PR

**Review by Mythie — APPROVED** [review]

**Comment by Mythie** [comment]
LGTM!

/tip 100

**Comment by tankerkiller125** [comment]
@Mythie thanks for the tip! It's appreciated.

**Comment by scoobs-mystery-bytes** [comment]
Anyone know what the environment variables would be for using this?

**Comment by tankerkiller125** [comment]
@NDCallahan Assuming it hasn't changed since this PR, 

```env
NEXT_PRIVATE_OIDC_WELL_KNOWN=""
NEXT_PRIVATE_OIDC_CLIENT_ID=""
NEXT_PRIVATE_OIDC_CLIENT_SECRET=""
```

Where the well known one points to the identity providers well-known OIDC URL.

**Comment by scoobs-mystery-bytes** [comment]
> @NDCallahan Assuming it hasn't changed since this PR,
> 
> ```dotenv
> NEXT_PRIVATE_OIDC_WELL_KNOWN=""
> NEXT_PRIVATE_OIDC_CLIENT_ID=""
> NEXT_PRIVATE_OIDC_CLIENT_SECRET=""
> ```
> 
> Where the well known one points to the identity providers well-known OIDC URL.

Thank you!

**Pull request merged**